src/o365/spo/commands/theme/theme-set.ts

```diff
@@ -0,0 +1,190 @@
+import SpoCommand from '../../SpoCommand';
+import { ContextInfo, ClientSvcResponse, ClientSvcResponseContents } from '../../spo';
+import * as fs from 'fs';
+import * as path from 'path';
+import Utils from '../../../../Utils';
+
+const vorpal: Vorpal = require('../../../../vorpal-init');
+
+interface CommandArgs {
+  options: Options;
+}
+
+interface Options extends GlobalOptions {
+  name: string;
+  filePath: string;
+  isInverted: boolean;
+}
+
+class ThemeSetCommand extends SpoCommand {
+
+  public get name(): string {
+    return commands.THEME_SET;
+  }
+
+  public get description(): string {
+    return 'Add or update theme to tenant with the given palette';
+  }
+
+  public getTelemetryProperties(args: CommandArgs): any {
+    const telemetryProps: any = super.getTelemetryProperties(args);
+    telemetryProps.inverted = (!(!args.options.isInverted)).toString();
+    return telemetryProps;
+  }
+
+  public commandAction(cmd: CommandInstance, args: CommandArgs, cb: () => void): void {
+
+    auth
+      .ensureAccessToken(auth.service.resource, cmd, this.debug)
+      .then((accessToken: string): request.RequestPromise => {
+        if (this.debug) {
+          cmd.log(`Retrieved access token ${accessToken}. Setting theme for the ${auth.site.url} tenant...`);
+        }
+
+        return this.getRequestDigest(cmd, this.debug);
+      })
+      .then((res: ContextInfo): request.RequestPromise => {
+        if (this.debug) {
+          cmd.log('Response:');
+          cmd.log(res);
+          cmd.log('');
+        }
+
+        const fullPath: string = path.resolve(args.options.filePath);
+
+        if (this.verbose) {
+          cmd.log(`Adding theme from ${fullPath} to tenant...`);
+        }
+
+        const palette: any = JSON.parse(fs.readFileSync(fullPath, 'utf8'));
+
+        if (this.debug) {
+          cmd.log('');
+          cmd.log('Palette');
+          cmd.log(JSON.stringify(palette));
+        }
+
+        const isInverted:boolean  = args.options.isInverted? true : false;
+
+        const requestOptions: any = {
+          url: `${auth.site.url}/_vti_bin/client.svc/ProcessQuery`,
+          headers: Utils.getRequestHeaders({
+            authorization: `Bearer ${auth.service.accessToken}`,
+            'X-RequestDigest': res.FormDigestValue
+          }),
+          body: `<Request AddExpandoFieldTypeSuffix="true" SchemaVersion="15.0.0.0" LibraryVersion="16.0.0.0" ApplicationName="${config.applicationName}" xmlns="http://schemas.microsoft.com/sharepoint/clientquery/2009"><Actions><ObjectPath Id="10" ObjectPathId="9" /><Method Name="UpdateTenantTheme" Id="11" ObjectPathId="9"><Parameters><Parameter Type="String">${args.options.name}</Parameter><Parameter Type="String">{"isInverted":${isInverted},"name":"${args.options.name}","palette":${JSON.stringify(palette)}}</Parameter></Parameters></Method></Actions><ObjectPaths><Constructor Id="9" TypeId="{268004ae-ef6b-4e9b-8425-127220d84719}"/></ObjectPaths></Request>`
```

Comment:
User input passed into the request should be XML-escaped. See for example: https://github.com/SharePoint/office365-cli/blob/42e0d29440fb857478129fdbfeddc4de7b9d4a9d/src/o365/spo/commands/storageentity/storageentity-set.ts#L80

Reply:
agreed. will make the changes!